tests/components/tasmota/test_device_trigger.py

```diff
@@ -0,0 +1,832 @@
+                    "action": {
+                        "service": "test.automation",
+                        "data_template": {"some": ("short_press")},
+                    },
+                },
+            ]
+        },
+    )
+
+    # Fake short press.
+    async_fire_mqtt_message(hass, "tasmota_49A3BC/stat/SWITCH1T", '{"TRIG":"TOGGLE"}')
+    await hass.async_block_till_done()
+    assert len(calls) == 1
+
+    # Update the trigger with different topic
+    async_fire_mqtt_message(hass, f"{DEFAULT_PREFIX}/{mac}/config", json.dumps(config2))
+    await hass.async_block_till_done()
+
+    async_fire_mqtt_message(hass, "tasmota_49A3BC/stat/SWITCH1T", '{"TRIG":"TOGGLE"}')
+    await hass.async_block_till_done()
+    assert len(calls) == 1
+
+    async_fire_mqtt_message(hass, "tasmota_49A3BC/status/SWITCH1T", '{"TRIG":"TOGGLE"}')
+    await hass.async_block_till_done()
+    assert len(calls) == 2
+
+    # Update the trigger with same topic
+    async_fire_mqtt_message(hass, f"{DEFAULT_PREFIX}/{mac}/config", json.dumps(config2))
+    await hass.async_block_till_done()
+
+    async_fire_mqtt_message(hass, "tasmota_49A3BC/stat/SWITCH1T", '{"TRIG":"TOGGLE"}')
+    await hass.async_block_till_done()
+    assert len(calls) == 2
+
+    async_fire_mqtt_message(hass, "tasmota_49A3BC/status/SWITCH1T", '{"TRIG":"TOGGLE"}')
+    await hass.async_block_till_done()
+    assert len(calls) == 3
+
+
+async def test_no_resubscribe_same_topic(hass, device_reg, mqtt_mock, setup_tasmota):
+    """Test subscription to topics without change."""
+    # Discover a device with device trigger
+    config = copy.deepcopy(DEFAULT_CONFIG)
+    config["swc"][0] = 0
+    mac = config["mac"]
+
+    mqtt_mock.async_subscribe.reset_mock()
+
+    async_fire_mqtt_message(hass, f"{DEFAULT_PREFIX}/{mac}/config", json.dumps(config))
+    await hass.async_block_till_done()
+
+    device_entry = device_reg.async_get_device(set(), {("mac", mac)})
+
+    assert await async_setup_component(
+        hass,
+        automation.DOMAIN,
+        {
+            automation.DOMAIN: [
+                {
+                    "trigger": {
+                        "platform": "device",
+                        "domain": DOMAIN,
+                        "device_id": device_entry.id,
+                        "discovery_id": "00000049A3BC_switch_1_TOGGLE",
+                        "type": "button_short_press",
+                        "subtype": "switch_1",
+                    },
+                    "action": {
+                        "service": "test.automation",
+                        "data_template": {"some": ("short_press")},
+                    },
+                },
+            ]
+        },
+    )
+
+    call_count = mqtt_mock.async_subscribe.call_count
+    assert call_count == 1
+    async_fire_mqtt_message(hass, f"{DEFAULT_PREFIX}/{mac}/config", json.dumps(config))
+    await hass.async_block_till_done()
+    assert mqtt_mock.async_subscribe.call_count == call_count
```

Comment:
Codacy found an issue: [Use of assert detected. The enclosed code will be removed when compiling to optimised byte code.](https://app.codacy.com/manual/foreign-sub/home-assistant/pullRequest?prid=6333224)